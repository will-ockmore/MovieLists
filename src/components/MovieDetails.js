import React, { Component } from 'react';
import { connect } from 'react-redux';

import { loadMovieDetails } from '../redux/actions';

import Loading from './Loading';


class MovieDetails extends Component {

  componentWillMount() {
    const { loadDetails, id } = this.props;
    loadDetails(id);
  }

  componentWillReceiveProps(nextProps) {
    const { loadDetails, id } = this.props;

    if (nextProps.params.id !== id) {
      loadDetails(nextProps.id);
    }
  }

  render() {
    const { details, largeBackdropUrl } = this.props
    const { result, status } = details.toObject();

    const {
      title, backdrop_path, overview,
      tagline, release_date, vote_average,
      original_title, credits, vote_count,
      runtime
    } = result.toObject();

    const { cast, crew } = credits.toObject();

    console.log(result.toJS());

    if (status === 'PENDING') {
      return <Loading />;
    }

    const hasMultipleVotes = !!vote_count && parseInt(vote_count, 10) > 1;

    return (
      <div className="movie-details">
        <img className="splash" src={largeBackdropUrl + backdrop_path} alt="poster" />
        <h1 className="mv-title">{title}{original_title && original_title !== title && ` (${original_title})`}</h1>
        {tagline && <h4 className="tagline">&ldquo;<em>{tagline}</em> &rdquo;</h4>}
        <div className="t-left mb-2">{overview}</div>

        {hasMultipleVotes &&
          <div className="fl-row">
            <div className="fl-1">
              <label className="property-title">
                <strong>Rating:</strong>
              </label>
            </div>
            <div className="fl-1">
              <label className="property-value">
                <strong>{vote_average}</strong>/10 &nbsp;<small>(from {vote_count} votes)</small>
              </label>
            </div>
          </div>
        }

        {runtime &&
          <div className="fl-row">
            <div className="fl-1">
              <label className="property-title"><strong>Runtime</strong>:</label>
            </div>
            <div className="fl-1">
              <label className="property-value">{runtime} mins</label>
            </div>
          </div>
        }

        {release_date &&
          <div className="fl-row">
            <div className="fl-1">
              <label className="property-title"><strong>Release date</strong>:</label>
            </div>
            <div className="fl-1">
              <label className="property-value">{release_date}</label>
            </div>
          </div>
        }

        <h2>Cast & crew</h2>

       {crew && crew.slice(0, 5).map(crewmember =>
          <div key={crewmember.get('credit_id')} className="fl-row">
            <div className="fl-1">
              <label className="property-title">
                <strong>{crewmember.get('job')}</strong>
              </label>
            </div>
            <div className="fl-1">
              <label className="property-value">{crewmember.get('name')}</label>
            </div>
          </div>
        )}

       <div className="mt-2">
         {cast && cast.slice(0, 15).map(castmember =>
            <div key={castmember.get('cast_id')} className="fl-row">
              <div className="fl-1">
                <label className="property-title">
                  <strong>{castmember.get('character')}</strong>
                </label>
              </div>
              <div className="fl-1">
                <label className="property-value">{castmember.get('name')}</label>
              </div>
            </div>
          )}
       </div>
      </div>
    );
  }
}

export function mapStateToProps(state, props) {
  const { id } = props.params;
  return {
    id,
    details: state.getIn(['responses', 'movieDetails']),
    largeBackdropUrl: state.getIn(['imageUrls', 'largeBackdropUrl'])
  }
}

export default connect(mapStateToProps, { loadDetails: loadMovieDetails })(MovieDetails);
