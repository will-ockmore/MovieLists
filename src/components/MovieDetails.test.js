import React from 'react';
import { Map } from 'immutable';
import renderer from 'react-test-renderer';

import { MOVIE_DETAILS } from '../../test/testFixtures';

import { MovieDetails, PropertyRow } from './MovieDetails';

describe('MovieDetails', () => {

  jest.mock('react-dom');

  it('renders the full tree', () => {
    const tree = renderer.create(
      <MovieDetails
        details={Map({result: MOVIE_DETAILS, status: 'COMPLETE'})}
        loadDetails={jest.fn}/>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders PropertyRow', () => {
    const tree = renderer.create(
      <PropertyRow propertyTitle="The title">
        <span>The children</span>
      </PropertyRow>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
