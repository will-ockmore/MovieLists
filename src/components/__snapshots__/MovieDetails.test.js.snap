exports[`MovieDetails renders PropertyRow 1`] = `
<div
  className="fl-row">
  <div
    className="fl-1">
    <label
      className="property-title">
      <strong>
        The title
      </strong>
    </label>
  </div>
  <div
    className="fl-1">
    <label
      className="property-value">
      <span>
        The children
      </span>
    </label>
  </div>
</div>
`;

exports[`MovieDetails renders the full tree 1`] = `
<div
  className="movie-details">
  <img
    alt="poster"
    className="splash"
    src="undefinedgood ruck kid" />
  <h1
    className="mv-title">
    The title
  </h1>
  <h4
    className="tagline">
    “
    <em>
      sick tagline bro
    </em>
     ”
  </h4>
  <div
    className="t-left mb-2">
    Some overview of the movie
  </div>
  <div
    className="fl-row">
    <div
      className="fl-1">
      <label
        className="property-title">
        <strong>
          Rating:
        </strong>
      </label>
    </div>
    <div
      className="fl-1">
      <label
        className="property-value">
        <strong>
          9.9
        </strong>
        /10  
        <small>
          (from 
          987
           votes)
        </small>
      </label>
    </div>
  </div>
  <div
    className="fl-row">
    <div
      className="fl-1">
      <label
        className="property-title">
        <strong>
          Release date:
        </strong>
      </label>
    </div>
    <div
      className="fl-1">
      <label
        className="property-value">
        1666-6-6
      </label>
    </div>
  </div>
  <h2>
    Cast & crew
  </h2>
  <div
    className="fl-row">
    <div
      className="fl-1">
      <label
        className="property-title">
        <strong>
          Camera
        </strong>
      </label>
    </div>
    <div
      className="fl-1">
      <label
        className="property-value">
        Chuck Norris
      </label>
    </div>
  </div>
  <div
    className="fl-row">
    <div
      className="fl-1">
      <label
        className="property-title">
        <strong>
          Lights
        </strong>
      </label>
    </div>
    <div
      className="fl-1">
      <label
        className="property-value">
        Chuck Norris
      </label>
    </div>
  </div>
  <div
    className="fl-row">
    <div
      className="fl-1">
      <label
        className="property-title">
        <strong>
          Action
        </strong>
      </label>
    </div>
    <div
      className="fl-1">
      <label
        className="property-value">
        Chuck Norris
      </label>
    </div>
  </div>
  <div
    className="mt-2">
    <div
      className="fl-row">
      <div
        className="fl-1">
        <label
          className="property-title">
          <strong>
            Sauron
          </strong>
        </label>
      </div>
      <div
        className="fl-1">
        <label
          className="property-value">
          The big boy
        </label>
      </div>
    </div>
    <div
      className="fl-row">
      <div
        className="fl-1">
        <label
          className="property-title">
          <strong>
            Gandalf
          </strong>
        </label>
      </div>
      <div
        className="fl-1">
        <label
          className="property-value">
          Ian McKellen
        </label>
      </div>
    </div>
    <div
      className="fl-row">
      <div
        className="fl-1">
        <label
          className="property-title">
          <strong>
            Frodo
          </strong>
        </label>
      </div>
      <div
        className="fl-1">
        <label
          className="property-value">
          Elijah Wood
        </label>
      </div>
    </div>
  </div>
</div>
`;
